File: core/src/test/java/com.memoizlabs.kotlin_playground/currency/ExchangeRatePresenterTest.java
package com.memoizlabs.kotlin_playground.currency;

import com.mvcoding.mvp.Presenter;

import org.junit.Before;
import org.junit.Test;

import rx.Observable;
import rx.subjects.PublishSubject;

import static com.memoizlabs.kotlin_playground.currency.GBPExchangeRateBuilder.aGBPExchangeRate;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ExchangeRatePresenterTest {

    ExchangeRatePresenter.View view = mock(ExchangeRatePresenter.View.class);
    GBPExchangeRateService gbpExchangeRateService = mock(GBPExchangeRateService.class);
    PublishSubject<Void> refreshes = PublishSubject.create();
    Presenter<ExchangeRatePresenter.View> presenter = new ExchangeRatePresenter(gbpExchangeRateService, new RxSchedulersForTest());

    GBPExchangeRate exchangeRate = aGBPExchangeRate().withEur(10).withUsd(5).build();

    @Before
    public void setup() {
        when(view.refreshes()).thenReturn(refreshes);
        when(gbpExchangeRateService.getRate()).thenReturn(Observable.just(exchangeRate));
    }

    @Test
    public void showsCurrency() {
        presenter.attach(view);

        refreshes.onNext(null);

        verify(view).showExchangeRate(exchangeRate);
    }
}

File: core/src/test/java/com.memoizlabs.kotlin_playground/currency/ExchangeRatePresenterTest.kt
package com.memoizlabs.kotlin_playground.currency

import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.*
import rx.Observable
import rx.subjects.PublishSubject

class ExchangeRatePresenterTest {

    private val view = mock(ExchangeRatePresenter.View::class.java)
    private val gbpExchangeRateService = mock(GBPExchangeRateService::class.java)
    private val refreshes = PublishSubject.create<Void>()
    private val presenter = ExchangeRatePresenter(gbpExchangeRateService, RxSchedulersForTest())

    private val exchangeRate = GBPExchangeRateBuilder.aGBPExchangeRate().withEur(10.0).withUsd(5.0).build()

    @Before
    fun setup() {
        `when`(view.refreshes()).thenReturn(refreshes)
        `when`(gbpExchangeRateService.getRate()).thenReturn(Observable.just(exchangeRate))
    }

    @Test
    fun showsCurrency() {
        presenter.attach(view)

        refreshes.onNext(null)

        verify(view).showExchangeRate(exchangeRate)
    }
}
